### DISCUSSION TITLE
Improving tests readability

### DISCUSSION THREAD

**Discussion by irega** [statelyai/xstate#2784] [Q&A] [answered]
Hi everyone!

We are using XState in our React Native application, and we are starting to implement the testing strategy. We are testing the state machine in [this way](https://xstate.js.org/docs/guides/testing.html#testing-services) as the official doc say, in order to test the side effects too and be able to mock some services we are using.

However, doing the tests in this way force us to use the `onTransition` callback at the beginning of the test, which we think is not the best for the readability and maintenance of the tests.

Do you know if it's possible to implement the tests in a more linear way? I mean, avoiding the callbacks and having a test structure like the [pure logic tests](https://xstate.js.org/docs/guides/testing.html#testing-pure-logic):

```typescript
// Given: a state machine and its mocked data/services
// When: we start the state machine and send an event
// Then: we check the expected state, context or mocked functions calls
```

Thank you so much!

**Comment by davidkpiano** [comment, marked as answer]
You can write a helper:

```js
function eventually(service, condition, timeout = 10000) {
  return new Promise((res, rej) => {
    setTimeout(rej, timeout);
    service.subscribe(state => {
      if (condition(state)) res();
    });
  });
}
```

And use it in your desired test structure:

```js
// Given
const machine = createMachine({ ... });

// When
const service = interpret(machine).start();
service.send({ type: 'SOME_EVENT' });

// Then
await eventually(service, state => state.matches('success'));
```

Is this what you mean?

**Reply by irega** [reply]
thank you @davidkpiano! Yes, that's we thought, encapsulate the code in an intermediate promise, but we were wondering if we missed something, and we were testing in the wrong way or there was some other function or utility from the library we could use.

**Reply by davidkpiano** [reply]
This is a good candidate for a helper function, and also something that @mattpocock and I have been discussing adding to core XState - the ability to "wait" for a condition to be true in a running service.
